### DISCUSSION TITLE
PDF Problems with Large Files

### DISCUSSION THREAD

**Discussion by skhilliard** [nextgenhealthcare/connect#4718] [Q&A] [answered]
Using Mirth 3.9.1

We are sending PDFs as base64 encoded strings in the Json body of an HTTP message to Mirth.

For testing, we forward these to a channel that writes them out as files using a file writer destination.  This file writer destination has a transformer that decodes the PDF contents and puts that value in a 'fileContent' channel map item (along with a 'fileName').  You can see in the image below how the 'fileContent' and 'fileName' items are used.

```
var input = $('inputJson')
var m = input.message;

var currentDate = DateUtil.getCurrentDate('yyyyMMddHHmmssssss');
var sentDate = DateUtil.convertDate('yyyyMMddHHmmss.ssss', 'yyyyMMddHHmm', m.messageTime);

if (m.combined) {
	var combined = new java.lang.String(FileUtil.decode(m.combined));
	channelMap.put('fileContent', '' + combined);
	channelMap.put('fileName', sentDate + '-' + m.lastName + '-' + m.firstName + currentDate + '.pdf');
}
```

The code above is hit when we are sending in a large "combined" PDF versus sending in the PDFs individually.  When we are dealing with the smaller "individual" PDFs, this is working just fine.  However, for the "combined" case the PDFs are corrupt/incomplete.  I suspect this is due some some kind of size limitation?  I took one of the "combined" base64 strings and decoded it myself, and it is valid.

Any thoughts/suggestions?

![Capture](https://user-images.githubusercontent.com/23263471/132061417-8c1117ca-5228-47c7-9ab0-bc329f89b085.PNG)

**Comment by pacmano1** [comment, marked as answer]
Kelly, this is Paul C from T, your former co-worker.  I am also "pacmano" over at the mirth forums.

I would probably use the attachment handler capability in Mirth for PDFs. see this article: https://forums.mirthproject.io/articles/175046-using-attachments

This functionality effectively lets you link to that attachment rather than have the large blob traverse your destinations.   

You should define "large" though, since in some work I do we regularly consume rather large CDAs and mirth handles that just fine, including normalized representations of the CDA.

**Reply by pacmano1** [reply]
So assuming your final destination is raw (and binary in the file writer):

Grab the string in the field "attachment" and store it as an attachment:

i.e.:
```
{
  "test" : "stuff",
  "attachment" : "<base 64 here>",
  "test1": "morestuff"
}
```




![forM](https://user-images.githubusercontent.com/44065187/132065880-115d85a2-696f-45a2-80f9-48f89a832e51.png)

In the destination transformer for a file writer:

![forM2](https://user-images.githubusercontent.com/44065187/132065758-c58d6368-b9fd-456f-9e2a-52a0b8cf555c.png)

And on the file writer:

![forM3](https://user-images.githubusercontent.com/44065187/132065821-121027c6-7e41-4c4c-9f9f-d2de3f45e994.png)

You can see the file in mirth:

![form4](https://user-images.githubusercontent.com/44065187/132065999-7068c55a-39e5-4a33-8cb1-625186eaf34d.png)

**Comment by skhilliard** [comment]
@pacmano1 Sorry for the late reply...got diverted onto other issues for a while.  I will try this out and let you know how it goes.

Kelly

**Reply by pacmano1** [reply]
@skhilliard Was this issue solved?  If yes, can you mark an response wth "Mark as Answer"?

**Comment by pacmano1** [comment]
Using attachments to reduce storage use was mentioned by OP at https://github.com/nextgenhealthcare/connect/discussions/4729
